// @forge/cli-shared/out/deploy/uploader/artifact-graphql-client.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const graphql_1 = require("../../graphql");
const ui_1 = require("../../ui");
class NoDeploymentURLError extends Error {
    constructor() {
        super(ui_1.Text.artifact.error.noDeployUrl);
    }
}
exports.NoDeploymentURLError = NoDeploymentURLError;
class ArtifactGraphQLClient {
    constructor(graphqlClient) {
        this.graphqlClient = graphqlClient;
    }
    async getUploadUrl(appDetails) {
        const artifactQuery = `
      mutation getUploadUrl($input: CreateAppDeploymentUrlInput!) {
        createAppDeploymentUrl(input: $input) {
          success
          message
          errorDetails {
            code
          }
          deploymentUrl
        }
      }
    `;
        const { response: { createAppDeploymentUrl: { deploymentUrl, success, errorDetails, message } }, requestId } = await this.graphqlClient.mutate(artifactQuery, {
            input: {
                appId: appDetails.id
            }
        });
        if (!success) {
            throw new graphql_1.GraphQlMutationError(`${message} (requestId: ${requestId || 'unknown'})`, errorDetails ? errorDetails.code : undefined);
        }
        if (!deploymentUrl) {
            throw new NoDeploymentURLError();
        }
        return deploymentUrl;
    }
}
exports.ArtifactGraphQLClient = ArtifactGraphQLClient;
