// @forge/cli-shared/out/file-system/file-system-reader.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const fs_1 = tslib_1.__importDefault(require("fs"));
const glob_1 = require("glob");
const os_1 = require("os");
const path_1 = tslib_1.__importDefault(require("path"));
const recursive_readdir_1 = tslib_1.__importDefault(require("recursive-readdir"));
const FILE_NOT_FOUND_CODE = 'ENOENT';
class FileSystemReader {
    fileExists(filePath) {
        return fs_1.default.existsSync(filePath);
    }
    readFile(filePath) {
        try {
            return fs_1.default.readFileSync(filePath, 'utf8');
        }
        catch (error) {
            return this.handleFileError(error);
        }
    }
    async recursiveReadDir(rootDir, ignores) {
        const ignoreDirs = ignores === null || ignores === void 0 ? void 0 : ignores.map((ignore) => {
            if (typeof ignore === 'string' && !path_1.default.extname(ignore)) {
                return (dir, stats) => stats.isDirectory() && path_1.default.basename(dir) === path_1.default.basename(ignore);
            }
            return ignore;
        });
        return recursive_readdir_1.default(rootDir, ignoreDirs);
    }
    async readGlob(glob) {
        return new Promise((resolve, reject) => {
            new glob_1.Glob(glob, {}, (err, matches) => {
                if (err) {
                    reject(err);
                }
                else {
                    resolve(matches);
                }
            });
        });
    }
    hasGlobSpecialChars(pattern) {
        return glob_1.hasMagic(pattern);
    }
    readBinaryFile(filePath) {
        try {
            return fs_1.default.readFileSync(filePath);
        }
        catch (error) {
            return this.handleFileError(error);
        }
    }
    getHomeDirectory() {
        return os_1.homedir();
    }
    getSize(filePath) {
        try {
            return fs_1.default.statSync(filePath).size;
        }
        catch (_a) {
            return undefined;
        }
    }
    handleFileError(error) {
        if (error && error.code === FILE_NOT_FOUND_CODE) {
            return undefined;
        }
        else {
            throw error;
        }
    }
}
exports.FileSystemReader = FileSystemReader;
