// @forge/lint/out/lint/linter-interface.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var LintClass;
(function (LintClass) {
    LintClass["Error"] = "error";
    LintClass["Warning"] = "warning";
})(LintClass = exports.LintClass || (exports.LintClass = {}));
class LintResult {
    constructor(file, fixer) {
        this.file = file;
        this.fixer = fixer;
        this.errors = [];
        this.warnings = [];
    }
    add(rule) {
        switch (rule.class) {
            case LintClass.Error:
                this.errors.push(rule);
                break;
            default:
                this.warnings.push(rule);
                break;
        }
    }
    batchAdd(...rules) {
        rules.forEach(this.add, this);
    }
    size() {
        return this.errors.length + this.warnings.length;
    }
    hasFixer() {
        return !!this.fixer;
    }
    async runFixer(state) {
        if (this.fixer) {
            return this.fixer(this.errors, this.warnings, state);
        }
        return state;
    }
}
exports.LintResult = LintResult;
