// @forge/cli-shared/out/apps/create-an-app.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const ui_1 = require("../ui");
class NoTemplatesError extends Error {
    constructor() {
        super(ui_1.Text.create.error.template.noTemplates);
    }
}
exports.NoTemplatesError = NoTemplatesError;
class CreateAppCommand {
    constructor(templater, registerAppCommand, packageInstaller) {
        this.templater = templater;
        this.registerAppCommand = registerAppCommand;
        this.packageInstaller = packageInstaller;
    }
    async execute({ name, description, template, directory }) {
        await this.templater.downloadAndExtract(template, directory);
        const appId = await this.registerAppCommand.registerApp({ name, description }, directory);
        await this.packageInstaller.install(directory);
        return { analytics: { template, appId, appName: name } };
    }
    async getAvailableTemplates() {
        const availableTemplates = await this.templater.getAvailableTemplates();
        if (!availableTemplates.length) {
            throw new NoTemplatesError();
        }
        return availableTemplates;
    }
}
exports.CreateAppCommand = CreateAppCommand;
