// @forge/cli-shared/out/apps/register-app.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const path_1 = require("path");
const ui_1 = require("../ui");
class RegisterAppCommand {
    constructor(appClient, appConfigReader, appConfigWriter, logger) {
        this.appClient = appClient;
        this.appConfigReader = appConfigReader;
        this.appConfigWriter = appConfigWriter;
        this.logger = logger;
    }
    async execute(input) {
        const appId = await this.registerApp(input, '.');
        return { analytics: { appId, appName: input.name } };
    }
    async isAppAlreadyConfigured() {
        const config = await this.appConfigReader.readConfigSection();
        if (config === 'missing' || config === undefined || config === 'invalid') {
            return 'missing';
        }
        return config.name;
    }
    async registerApp(input, directory) {
        this.logger.info(ui_1.Text.create.taskRegister);
        const id = await this.appClient.createApp(input);
        const cwd = process.cwd();
        process.chdir(path_1.resolve(cwd, directory));
        try {
            await this.appConfigWriter.writeConfigSection(Object.assign({ id }, input));
        }
        finally {
            process.chdir(cwd);
        }
        this.logger.info(ui_1.Text.create.taskCreateEnvironments);
        return id;
    }
}
exports.RegisterAppCommand = RegisterAppCommand;
