// @forge/cli-shared/out/ui/log-color.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const chalk_1 = tslib_1.__importDefault(require("chalk"));
exports.LogColor = {
    trace: (message) => chalk_1.default.dim(message),
    debug: (message) => chalk_1.default.gray(message),
    warn: (message) => chalk_1.default.yellow(message),
    error: (message) => chalk_1.default.red(message),
    errorPrefix: chalk_1.default.red.bold('Error: ')
};
const withPadding = (input) => input.padEnd(7);
exports.RuntimeLogLevel = {
    trace: chalk_1.default.dim(withPadding('TRACE')),
    debug: chalk_1.default.magenta(withPadding('DEBUG')),
    info: chalk_1.default.cyan(withPadding('INFO')),
    warn: chalk_1.default.yellow(withPadding('WARN')),
    error: chalk_1.default.red(withPadding('ERROR')),
    fatal: chalk_1.default.bgRedBright.bold.hex('#fff')(withPadding('FATAL'))
};
exports.TableStyle = {
    default: {
        style: { head: [] }
    },
    runtimeLog: {
        style: { head: ['gray'], border: ['gray'] },
        chars: {
            mid: '',
            middle: '',
            'left-mid': '',
            'mid-mid': '',
            'right-mid': '',
            'top-mid': '',
            'bottom-mid': ''
        }
    }
};
