// @forge/lint/out/parse/babel-parser.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const parser_interface_1 = require("./parser-interface");
const core_1 = require("@babel/core");
const DEFAULT_BABEL_CONFIG = {
    presets: [
        [
            require.resolve('@babel/preset-typescript'),
            {
                isTSX: true,
                jsxPragma: 'ForgeUI',
                allExtensions: true
            }
        ]
    ],
    plugins: [
        [
            require.resolve('@babel/plugin-transform-react-jsx'),
            {
                pragma: 'ForgeUI.createElement'
            }
        ],
        require.resolve('@forge/babel-plugin-transform-ui'),
        require.resolve('@babel/plugin-proposal-numeric-separator'),
        require.resolve('@babel/plugin-proposal-class-properties')
    ]
};
exports.babelParser = async (code) => {
    const parsed = await core_1.parseAsync(code, DEFAULT_BABEL_CONFIG);
    return {
        type: parser_interface_1.ParserTypes.Babel,
        parsed,
        traverse: (options) => {
            if (parsed) {
                core_1.traverse(parsed, options);
            }
        }
    };
};
