// @forge/cli-shared/out/http-client/global-edge-http-client.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const url_1 = require("url");
const ui_1 = require("../ui");
class InvalidSiteError extends Error {
    constructor(url) {
        super(ui_1.Text.install.error.url(url.toString()));
    }
}
exports.InvalidSiteError = InvalidSiteError;
class InvalidResponseError extends Error {
    constructor(url) {
        super(ui_1.Text.install.error.site(url.toString()));
    }
}
exports.InvalidResponseError = InvalidResponseError;
class GlobalEdgeHttpClient {
    constructor(fetchClient) {
        this.fetchClient = fetchClient;
    }
    async getCloudId(site) {
        const url = this.buildUrl(site);
        const res = await this.fetchClient.fetch(url);
        if (!res.ok) {
            throw new InvalidSiteError(site);
        }
        let json;
        try {
            json = await res.json();
        }
        catch (err) {
            throw new InvalidResponseError(site);
        }
        if (json && json.cloudId) {
            return json.cloudId;
        }
        throw new InvalidResponseError(site);
    }
    buildUrl(site) {
        const urlObj = new url_1.URL('/_edge/tenant_info', site.toString());
        return urlObj.toString();
    }
}
exports.GlobalEdgeHttpClient = GlobalEdgeHttpClient;
