// @forge/cli-shared/out/auth/container/token-graphql-client.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const graphql_1 = require("../../graphql");
class CannotCreateContainerToken extends Error {
}
exports.CannotCreateContainerToken = CannotCreateContainerToken;
class ContainerTokenCreatingGraphqlClient {
    constructor(graphqlClient) {
        this.graphqlClient = graphqlClient;
    }
    async createToken({ appId }) {
        const query = `
      mutation createContainerToken($input: CreateContainerTokenInput!) {
        createContainerToken(input: $input) {
          message
          success
          errorDetails {
            code
          }
          token
        }
      }
    `;
        const variables = {
            input: {
                appId
            }
        };
        const { response: { createContainerToken: { token, message, success, errorDetails } }, requestId } = await this.graphqlClient.mutate(query, variables);
        if (!success) {
            throw new graphql_1.GraphQlMutationError(`${message} (requestId: ${requestId || 'unknown'})`, errorDetails ? errorDetails.code : undefined);
        }
        if (!token) {
            throw new CannotCreateContainerToken(message);
        }
        return token;
    }
}
exports.ContainerTokenCreatingGraphqlClient = ContainerTokenCreatingGraphqlClient;
