// @forge/cli-shared/out/auth/personal/abstract-token-store-getter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const instruction_1 = require("../instruction");
const ui_1 = require("../../ui");
exports.EMAIL_KEY = 'FORGE_EMAIL';
exports.API_TOKEN_KEY = 'FORGE_API_TOKEN';
exports.AAID_KEY = 'FORGE_AAID';
class AbstractPersonalApiCredentialsGetter {
    constructor(graphqlGateway) {
        this.graphqlGateway = graphqlGateway;
        this.getCredentials = () => {
            if (process.env[exports.API_TOKEN_KEY] && process.env[exports.EMAIL_KEY]) {
                return Promise.resolve({
                    email: process.env[exports.EMAIL_KEY],
                    token: process.env[exports.API_TOKEN_KEY],
                    accountId: process.env[exports.AAID_KEY] || ''
                });
            }
            return this.getCredentialsImpl();
        };
        this.getInstructionsWhenInvalid = () => {
            if (process.env[exports.API_TOKEN_KEY] && process.env[exports.EMAIL_KEY]) {
                const instructionsUrl = instruction_1.getInstructionsUrl(this.graphqlGateway);
                return ui_1.Text.error.invalidTokenInEnv(instructionsUrl, exports.EMAIL_KEY, exports.API_TOKEN_KEY);
            }
            return this.getInstructionsWhenInvalidImpl();
        };
    }
}
exports.AbstractPersonalApiCredentialsGetter = AbstractPersonalApiCredentialsGetter;
