// @forge/cli-shared/out/ari/ari.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const cs_ari_1 = require("@forge/util/packages/cs-ari");
function appIdToAriString(appId) {
    return `ari:cloud:ecosystem::app/${appId}`;
}
exports.appIdToAriString = appIdToAriString;
function buildContextAri(cloudId, product) {
    return new cs_ari_1.ResourceIdentifier({
        resourceOwner: product.toLowerCase(),
        resourceType: 'site',
        resourceId: cloudId
    });
}
exports.buildContextAri = buildContextAri;
function buildExtensionAri(appId, environmentId, functionKey) {
    return new cs_ari_1.ResourceIdentifier({
        resourceOwner: 'ecosystem',
        resourceType: 'extension',
        resourceId: `${appId}/${environmentId}/static/${functionKey}`
    });
}
exports.buildExtensionAri = buildExtensionAri;
function encodeContext(context, extension) {
    const serialised = JSON.stringify({ ctx: context.toString(), ext: extension.toString() });
    return Buffer.from(serialised, 'utf-8')
        .toString('base64')
        .replace(/=/g, '')
        .replace(/\+/g, '-')
        .replace(/\//g, '_');
}
exports.encodeContext = encodeContext;
