// @forge/api/out/store/global/global-storage.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const errors_1 = require("../errors");
const queries_1 = require("./queries");
class GlobalStorage {
    constructor(apiClient) {
        this.apiClient = apiClient;
        this.endpoint = '/forge/entities/graphql';
    }
    getAppContextAri() {
        var _a, _b;
        if (!this.appContextAri) {
            this.appContextAri = (_b = (_a = global.api).__getAppAri) === null || _b === void 0 ? void 0 : _b.call(_a);
        }
        return this.appContextAri;
    }
    async get(key) {
        const requestBody = queries_1.getQuery(this.getAppContextAri(), key);
        const { appStoredEntity: { value } } = await this.query(requestBody);
        return value || undefined;
    }
    async list(options) {
        const requestBody = queries_1.listQuery(this.getAppContextAri(), options);
        const { appStoredEntities: { edges } } = await this.query(requestBody);
        const nextCursor = edges.length > 0 ? edges[edges.length - 1].cursor : undefined;
        const results = edges.map(({ node }) => node);
        return {
            results,
            nextCursor
        };
    }
    async set(key, value) {
        const requestBody = queries_1.setQuery(this.getAppContextAri(), key, value);
        await this.mutation(requestBody, 'setAppStoredEntity');
    }
    async delete(key) {
        const requestBody = queries_1.deleteQuery(this.getAppContextAri(), key);
        await this.mutation(requestBody, 'deleteAppStoredEntity');
    }
    buildRequest(requestBody) {
        return {
            method: 'POST',
            body: JSON.stringify(requestBody),
            headers: {
                'content-type': 'application/json'
            }
        };
    }
    async query(body) {
        const response = await this.apiClient(this.endpoint, this.buildRequest(body));
        if (!response.ok) {
            throw new errors_1.StorageAPIError(response.status);
        }
        const { errors, data } = await response.json();
        if (errors) {
            const { statusCode } = errors[0].extensions;
            throw new errors_1.StorageAPIError(statusCode);
        }
        return data;
    }
    async mutation(body, mutationMethod) {
        const response = await this.apiClient(this.endpoint, this.buildRequest(body));
        if (!response.ok) {
            throw new errors_1.StorageAPIError(response.status);
        }
        const { data: { appStorage: { [mutationMethod]: { success, errors } } } } = await response.json();
        if (errors && errors.length > 0) {
            const { statusCode } = errors[0].extensions;
            throw new errors_1.StorageAPIError(statusCode);
        }
        if (!success) {
            throw new errors_1.StorageAPIError(500);
        }
        return response;
    }
}
exports.GlobalStorage = GlobalStorage;
