// @forge/cli-shared/out/graphql/app-environment-graphql-client.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const ui_1 = require("../ui");
class MissingAppError extends Error {
    constructor() {
        super(ui_1.Text.env.error.appNotExist);
    }
}
exports.MissingAppError = MissingAppError;
class MissingAppEnvironmentError extends Error {
    constructor() {
        super(ui_1.Text.env.error.envNotExist);
    }
}
exports.MissingAppEnvironmentError = MissingAppEnvironmentError;
class AppEnvironmentsGraphqlClient {
    constructor(graphqlClient) {
        this.graphqlClient = graphqlClient;
    }
    async getAppEnvironmentId(appId, environmentKey) {
        const query = `
      query app($id: ID!, $key: String!) {
        app(id: $id) {
          environmentByKey(key: $key) {
            id
          }
        }
      }
    `;
        const result = await this.graphqlClient.query(query, {
            id: appId,
            key: environmentKey
        });
        if (!result.app) {
            throw new MissingAppError();
        }
        if (!result.app.environmentByKey) {
            throw new MissingAppEnvironmentError();
        }
        return result.app.environmentByKey.id;
    }
}
exports.AppEnvironmentsGraphqlClient = AppEnvironmentsGraphqlClient;
