// @forge/cli-shared/out/graphql/graphql-types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var AccountStatus;
(function (AccountStatus) {
    AccountStatus["Active"] = "active";
    AccountStatus["Inactive"] = "inactive";
    AccountStatus["Closed"] = "closed";
})(AccountStatus = exports.AccountStatus || (exports.AccountStatus = {}));
var ActivitiesContainerType;
(function (ActivitiesContainerType) {
    ActivitiesContainerType["Site"] = "SITE";
    ActivitiesContainerType["Project"] = "PROJECT";
    ActivitiesContainerType["Space"] = "SPACE";
})(ActivitiesContainerType = exports.ActivitiesContainerType || (exports.ActivitiesContainerType = {}));
var ActivitiesFilterType;
(function (ActivitiesFilterType) {
    ActivitiesFilterType["And"] = "AND";
    ActivitiesFilterType["Or"] = "OR";
})(ActivitiesFilterType = exports.ActivitiesFilterType || (exports.ActivitiesFilterType = {}));
var ActivitiesObjectType;
(function (ActivitiesObjectType) {
    ActivitiesObjectType["Issue"] = "ISSUE";
    ActivitiesObjectType["Page"] = "PAGE";
    ActivitiesObjectType["Blogpost"] = "BLOGPOST";
})(ActivitiesObjectType = exports.ActivitiesObjectType || (exports.ActivitiesObjectType = {}));
var ActivityEventType;
(function (ActivityEventType) {
    ActivityEventType["Assigned"] = "ASSIGNED";
    ActivityEventType["Unassigned"] = "UNASSIGNED";
    ActivityEventType["Viewed"] = "VIEWED";
    ActivityEventType["Commented"] = "COMMENTED";
    ActivityEventType["Updated"] = "UPDATED";
    ActivityEventType["Created"] = "CREATED";
    ActivityEventType["Liked"] = "LIKED";
    ActivityEventType["Transitioned"] = "TRANSITIONED";
    ActivityEventType["Published"] = "PUBLISHED";
    ActivityEventType["Edited"] = "EDITED";
})(ActivityEventType = exports.ActivityEventType || (exports.ActivityEventType = {}));
var ActivityObjectType;
(function (ActivityObjectType) {
    ActivityObjectType["Site"] = "SITE";
    ActivityObjectType["Project"] = "PROJECT";
    ActivityObjectType["Issue"] = "ISSUE";
    ActivityObjectType["Comment"] = "COMMENT";
    ActivityObjectType["Space"] = "SPACE";
    ActivityObjectType["Page"] = "PAGE";
    ActivityObjectType["Blogpost"] = "BLOGPOST";
    ActivityObjectType["Task"] = "TASK";
})(ActivityObjectType = exports.ActivityObjectType || (exports.ActivityObjectType = {}));
var ActivityProduct;
(function (ActivityProduct) {
    ActivityProduct["Jira"] = "JIRA";
    ActivityProduct["JiraBusiness"] = "JIRA_BUSINESS";
    ActivityProduct["JiraSoftware"] = "JIRA_SOFTWARE";
    ActivityProduct["JiraOps"] = "JIRA_OPS";
    ActivityProduct["JiraServiceDesk"] = "JIRA_SERVICE_DESK";
    ActivityProduct["Confluence"] = "CONFLUENCE";
})(ActivityProduct = exports.ActivityProduct || (exports.ActivityProduct = {}));
var AppDeploymentEventLogLevel;
(function (AppDeploymentEventLogLevel) {
    AppDeploymentEventLogLevel["Info"] = "INFO";
    AppDeploymentEventLogLevel["Warning"] = "WARNING";
    AppDeploymentEventLogLevel["Error"] = "ERROR";
})(AppDeploymentEventLogLevel = exports.AppDeploymentEventLogLevel || (exports.AppDeploymentEventLogLevel = {}));
var AppDeploymentStatus;
(function (AppDeploymentStatus) {
    AppDeploymentStatus["InProgress"] = "IN_PROGRESS";
    AppDeploymentStatus["Done"] = "DONE";
    AppDeploymentStatus["Failed"] = "FAILED";
})(AppDeploymentStatus = exports.AppDeploymentStatus || (exports.AppDeploymentStatus = {}));
var AppDeploymentStepStatus;
(function (AppDeploymentStepStatus) {
    AppDeploymentStepStatus["Started"] = "STARTED";
    AppDeploymentStepStatus["Done"] = "DONE";
    AppDeploymentStepStatus["Failed"] = "FAILED";
})(AppDeploymentStepStatus = exports.AppDeploymentStepStatus || (exports.AppDeploymentStepStatus = {}));
var AppEnvironmentType;
(function (AppEnvironmentType) {
    AppEnvironmentType["Testing"] = "TESTING";
    AppEnvironmentType["Development"] = "DEVELOPMENT";
    AppEnvironmentType["Production"] = "PRODUCTION";
})(AppEnvironmentType = exports.AppEnvironmentType || (exports.AppEnvironmentType = {}));
var AppStoredEntityCondition;
(function (AppStoredEntityCondition) {
    AppStoredEntityCondition["StartsWith"] = "STARTS_WITH";
})(AppStoredEntityCondition = exports.AppStoredEntityCondition || (exports.AppStoredEntityCondition = {}));
var BoardFeatureStatus;
(function (BoardFeatureStatus) {
    BoardFeatureStatus["Enabled"] = "ENABLED";
    BoardFeatureStatus["Disabled"] = "DISABLED";
    BoardFeatureStatus["ComingSoon"] = "COMING_SOON";
})(BoardFeatureStatus = exports.BoardFeatureStatus || (exports.BoardFeatureStatus = {}));
var BoardFeatureToggleStatus;
(function (BoardFeatureToggleStatus) {
    BoardFeatureToggleStatus["Enabled"] = "ENABLED";
    BoardFeatureToggleStatus["Disabled"] = "DISABLED";
})(BoardFeatureToggleStatus = exports.BoardFeatureToggleStatus || (exports.BoardFeatureToggleStatus = {}));
var DevStatusActivity;
(function (DevStatusActivity) {
    DevStatusActivity["PrOpen"] = "PR_OPEN";
    DevStatusActivity["PrMerged"] = "PR_MERGED";
    DevStatusActivity["PrDeclined"] = "PR_DECLINED";
    DevStatusActivity["BranchOpen"] = "BRANCH_OPEN";
    DevStatusActivity["Commit"] = "COMMIT";
})(DevStatusActivity = exports.DevStatusActivity || (exports.DevStatusActivity = {}));
var FunctionTriggerType;
(function (FunctionTriggerType) {
    FunctionTriggerType["Frontend"] = "FRONTEND";
    FunctionTriggerType["Manual"] = "MANUAL";
    FunctionTriggerType["Product"] = "PRODUCT";
    FunctionTriggerType["Web"] = "WEB";
})(FunctionTriggerType = exports.FunctionTriggerType || (exports.FunctionTriggerType = {}));
var IssueDevOpsCommitChangeType;
(function (IssueDevOpsCommitChangeType) {
    IssueDevOpsCommitChangeType["Added"] = "ADDED";
    IssueDevOpsCommitChangeType["Deleted"] = "DELETED";
    IssueDevOpsCommitChangeType["Copied"] = "COPIED";
    IssueDevOpsCommitChangeType["Moved"] = "MOVED";
    IssueDevOpsCommitChangeType["Modify"] = "MODIFY";
    IssueDevOpsCommitChangeType["Unknown"] = "UNKNOWN";
})(IssueDevOpsCommitChangeType = exports.IssueDevOpsCommitChangeType || (exports.IssueDevOpsCommitChangeType = {}));
var IssueDevOpsDeploymentEnvironmentType;
(function (IssueDevOpsDeploymentEnvironmentType) {
    IssueDevOpsDeploymentEnvironmentType["Production"] = "PRODUCTION";
    IssueDevOpsDeploymentEnvironmentType["Staging"] = "STAGING";
    IssueDevOpsDeploymentEnvironmentType["Testing"] = "TESTING";
    IssueDevOpsDeploymentEnvironmentType["Development"] = "DEVELOPMENT";
    IssueDevOpsDeploymentEnvironmentType["Unmapped"] = "UNMAPPED";
})(IssueDevOpsDeploymentEnvironmentType = exports.IssueDevOpsDeploymentEnvironmentType || (exports.IssueDevOpsDeploymentEnvironmentType = {}));
var IssueDevOpsDeploymentState;
(function (IssueDevOpsDeploymentState) {
    IssueDevOpsDeploymentState["Pending"] = "PENDING";
    IssueDevOpsDeploymentState["InProgress"] = "IN_PROGRESS";
    IssueDevOpsDeploymentState["Successful"] = "SUCCESSFUL";
    IssueDevOpsDeploymentState["Cancelled"] = "CANCELLED";
    IssueDevOpsDeploymentState["Failed"] = "FAILED";
    IssueDevOpsDeploymentState["RolledBack"] = "ROLLED_BACK";
    IssueDevOpsDeploymentState["Unknown"] = "UNKNOWN";
})(IssueDevOpsDeploymentState = exports.IssueDevOpsDeploymentState || (exports.IssueDevOpsDeploymentState = {}));
var IssueDevOpsPullRequestStatus;
(function (IssueDevOpsPullRequestStatus) {
    IssueDevOpsPullRequestStatus["Open"] = "OPEN";
    IssueDevOpsPullRequestStatus["Merged"] = "MERGED";
    IssueDevOpsPullRequestStatus["Declined"] = "DECLINED";
})(IssueDevOpsPullRequestStatus = exports.IssueDevOpsPullRequestStatus || (exports.IssueDevOpsPullRequestStatus = {}));
var Product;
(function (Product) {
    Product["Jira"] = "Jira";
    Product["Confluence"] = "Confluence";
})(Product = exports.Product || (exports.Product = {}));
var SoftwareCardsDestinationEnum;
(function (SoftwareCardsDestinationEnum) {
    SoftwareCardsDestinationEnum["NewSprint"] = "NEW_SPRINT";
    SoftwareCardsDestinationEnum["Backlog"] = "BACKLOG";
    SoftwareCardsDestinationEnum["ExistingSprint"] = "EXISTING_SPRINT";
})(SoftwareCardsDestinationEnum = exports.SoftwareCardsDestinationEnum || (exports.SoftwareCardsDestinationEnum = {}));
var SprintReportsEstimationStatisticType;
(function (SprintReportsEstimationStatisticType) {
    SprintReportsEstimationStatisticType["StoryPoints"] = "STORY_POINTS";
    SprintReportsEstimationStatisticType["IssueCount"] = "ISSUE_COUNT";
})(SprintReportsEstimationStatisticType = exports.SprintReportsEstimationStatisticType || (exports.SprintReportsEstimationStatisticType = {}));
var SprintState;
(function (SprintState) {
    SprintState["Active"] = "ACTIVE";
    SprintState["Future"] = "FUTURE";
    SprintState["Closed"] = "CLOSED";
})(SprintState = exports.SprintState || (exports.SprintState = {}));
var SwimlaneStrategy;
(function (SwimlaneStrategy) {
    SwimlaneStrategy["None"] = "NONE";
    SwimlaneStrategy["Assignee"] = "ASSIGNEE";
    SwimlaneStrategy["Issueparent"] = "ISSUEPARENT";
    SwimlaneStrategy["Issuechildren"] = "ISSUECHILDREN";
})(SwimlaneStrategy = exports.SwimlaneStrategy || (exports.SwimlaneStrategy = {}));
